### PR TITLE
Separate into load test creator and runner

### PR DESCRIPTION
**Pull request by ajmaddaford**
The current survey runner benchmark takes a generated har file and uses it to run a load test. This isn't very flexible as it means modifying the test after the initial capture is difficult and error prone. This pull request introduces an intermediary json file that contains a subset of the HAR file required to run a load test. These files can be checked in and modified as and when the schema changes or we want to create new scenarios.

## How to review
- Generate a JSON requests file from a HAR file
- Run a load test using the JSON requests file

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,64 +1,59 @@
...
+  - Defaults to `requests.json`
 - `HOST` - The host name the benchmark is run against.
 - `USER_WAIT_TIME_MIN_SECONDS` - The minimum delay between each user's GET requests
   - defaults to zero
 - `USER_WAIT_TIME_MAX_SECONDS` - The maximum delay between each user's GET requests
   - defaults to zero
 
+## Generating a requests file
+
+Open the network inspector in Chrome or Firefox and ensure 'preserve log' is ticked. Run your manual or functional test.
```
Comment (bitdivision) [inline README.md]: Would be good to include the fact that the session should not be included here. That's caused some confusion in the past.

**Comment by bitdivision** [comment]
I'm running into the following error:
```Traceback (most recent call last):
  File "generate_requests.py", line 63, in <module>
    generate_requests()
  File "/home/bitdivision/.local/share/virtualenvs/eq-survey-runner-benchmark-iJrwGlGv/lib/python3.7/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/home/bitdivision/.local/share/virtualenvs/eq-survey-runner-benchmark-iJrwGlGv/lib/python3.7/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/home/bitdivision/.local/share/virtualenvs/eq-survey-runner-benchmark-iJrwGlGv/lib/python3.7/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/bitdivision/.local/share/virtualenvs/eq-survey-runner-benchmark-iJrwGlGv/lib/python3.7/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "generate_requests.py", line 57, in generate_requests
    requests = parse_har_file(har_file)
  File "generate_requests.py", line 20, in parse_har_file
    for page in har_parser.pages:
  File "/home/bitdivision/.local/share/virtualenvs/eq-survey-runner-benchmark-iJrwGlGv/lib/python3.7/site-packages/haralyzer/assets.py", line 185, in pages
    page = HarPage(har_page['id'], har_parser=self)
  File "/home/bitdivision/.local/share/virtualenvs/eq-survey-runner-benchmark-iJrwGlGv/lib/python3.7/site-packages/haralyzer/assets.py", line 246, in __init__
    self.title = page['title']
KeyError: 'title'
```

That may be an idiosyncrasy of firefox HAR files though. Will try again in chrome

**Comment by bitdivision** [comment]
Works fine in chrome

File: generate_requests.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
+import json
+from operator import itemgetter
+from pathlib import Path
```
Comment (MebinAbraham) [inline generate_requests.py]: Can we remove the unused imports from this file, please?

File: runner_benchmark/taskset.py [hunk 3]
```diff
@@ -1,3 +1,4 @@
+import json
```
Comment (MebinAbraham) [inline runner_benchmark/taskset.py]: Can we remove the unused imports from this file, please?

File: runner_benchmark/taskset.py [hunk 4]
```diff
@@ -39,43 +36,47 @@ def replay_requests(self):
...
                     time.sleep(r.randrange(user_wait_time_min, user_wait_time_max))
-            elif request['method'] == 'POST':
 
+            elif request['method'] == 'POST':
                 print('POST to ', request['url'])
 
-                response = self.post(**request)
+                response = self.post(request['url'], request['data'])
 
                 if response.status_code != 302:
```
Comment (MebinAbraham) [inline runner_benchmark/taskset.py]: This needs to accept a `200` also as a `POST` with an invalid form returns a `200` response.

File: runner_benchmark/taskset.py [hunk 5]
```diff
@@ -39,43 +36,47 @@ def replay_requests(self):
...
-                    raise Exception(f"Got a non-302 ({response.status_code}) back when posting page: {request['url']} with data: {request['data']}")
+                    raise Exception(
+                        f"Got a non-302 ({response.status_code}) back when posting page: {request['url']} with data: {request['data']}"
+                    )
 
                 print(f"Redirect to: {response.headers['Location']}")
```
Comment (MebinAbraham) [inline runner_benchmark/taskset.py]: Would currently give a KeyError when return response is 200.

File: runner_benchmark/taskset.py [hunk 6]
```diff
@@ -39,43 +36,47 @@ def replay_requests(self):
...
         if response.status_code != 302:
-            raise Exception('Got a non-302 back when authenticating session: {}'.format(response.status_code))
+            raise Exception(
+                'Got a non-302 back when authenticating session: {}'.format(
+                    response.status_code
+                )
+            )
 
-        return self.get(url=response.headers['Location'])
+        response = self.get(response.headers['Location'], allow_redirects=True)
```
Comment (MebinAbraham) [inline runner_benchmark/taskset.py]: How come this is assigned to a variable?

Comment (ajmaddaford) [reply runner_benchmark/taskset.py]: Originally I was using the response. I've removed the variable.

**Pull request merged**